src/_types.ts

```diff
@@ -12,6 +12,20 @@ export type PredIndexedOptional<T, K> = (
   array?: ReadonlyArray<T>,
 ) => K;
 
+export type TypePred<T, S extends T> = (input: T) => input is S;
+
+export type TypePredIndexed<T, S extends T> = (
+  input: T,
+  index: number,
+  array: ReadonlyArray<T>,
+) => input is S;
+
+export type TypePredIndexedOptional<T, S extends T> = (
+  input: T,
+  index?: number,
+  array?: ReadonlyArray<T>,
+) => input is S;
```

Comment:
Don't define these here, all of these types are going away in v2. They aren't useful because the hide a very basic core concept of typing

Reply:
Where would be a better place to put them then, while in v1?

Reply:
Put them inline in the file, the reasoning behind putting them in `_types` was that I immediately needed the same thing for `some` in the next PR, but let's take it step by step

Reply:
@eranhirsch lgty?